Enderpearl bow projectile error

Read: https://github.com/Electroid/PGM/issues/57#issuecomment-551205212

~~Add custom projectiles so maps like Ozone works properly, and so we can have cool snowball bows on Glacial Impact;~~

```xml
<modifybowprojectile>
    <projectile>EnderPearl</projectile>
    <velocityMod>2.5</velocityMod>
</modifybowprojectile>
```
```xml
<modifybowprojectile>
    <projectile>Snowball</projectile>
    <velocityMod>2.0</velocityMod>
</modifybowprojectile>
```

Ah okay. Well, it works on glacial impact as intended now, but not on Ozone.

```java
[19:28:49 ERROR]: Could not pass event EntityShootBowEvent to PGM v1.8-SNAPSHOT
org.bukkit.event.EventException
        at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:302) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:74) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:65) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.callEventHandler(SimplePluginManager.java:569) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.dispatchEvent(SimplePluginManager.java:534) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.access$0(SimplePluginManager.java:531) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager$2.callEvent(SimplePluginManager.java:536) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.callEventBody(SimplePluginManager.java:586) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.callEventHandler(SimplePluginManager.java:579) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.dispatchEvent(SimplePluginManager.java:534) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.callEventSynchronously(SimplePluginManager.java:527) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:521) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:496) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.craftbukkit.v1_8_R3.event.CraftEventFactory.callEntityShootBowEvent(CraftEventFactory.java:250) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.ItemBow.a(ItemBow.java:61) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.ItemStack.b(ItemStack.java:527) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.EntityHuman.bU(EntityHuman.java:109) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.PlayerConnection.a(PlayerConnection.java:631) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.PacketPlayInBlockDig.a(SourceFile:40) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.PacketPlayInBlockDig.a(SourceFile:10) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.PlayerConnectionUtils$1.run(SourceFile:13) [sportpaper.jar:git-SportPaper-"374af6d"]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_222]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_222]
        at net.minecraft.server.v1_8_R3.SystemUtils.a(SourceFile:44) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.MinecraftServer.processTasks(MinecraftServer.java:1724) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.DedicatedServer.processTasks(DedicatedServer.java:397) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.MinecraftServer.B(MinecraftServer.java:882) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.DedicatedServer.B(DedicatedServer.java:365) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.MinecraftServer.A(MinecraftServer.java:821) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.MinecraftServer.run(MinecraftServer.java:722) [sportpaper.jar:git-SportPaper-"374af6d"]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_222]
Caused by: java.lang.NullPointerException
        at net.minecraft.server.v1_8_R3.EntityProjectile.<init>(EntityProjectile.java:31) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.EntityEnderPearl.<init>(EntityEnderPearl.java:19) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.craftbukkit.v1_8_R3.CraftWorld.createEntity(CraftWorld.java:1104) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.craftbukkit.v1_8_R3.CraftWorld.spawn(CraftWorld.java:1319) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.craftbukkit.v1_8_R3.CraftWorld.spawn(CraftWorld.java:1034) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        at tc.oc.pgm.modules.ModifyBowProjectileMatchModule.changeBowProjectile(ModifyBowProjectileMatchModule.java:53) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_222]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_222]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_222]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_222]
        at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:300) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        ... 30 more
```

```xml
<modifybowprojectile>
    <projectile>EnderPearl</projectile>
    <velocityMod>2.5</velocityMod>
</modifybowprojectile>
```

--- a/src/main/java/tc/oc/pgm/modules/ModifyBowProjectileMatchModule.java
+++ b/src/main/java/tc/oc/pgm/modules/ModifyBowProjectileMatchModule.java
@@ -2,11 +2,7 @@ package tc.oc.pgm.modules;
 
 import java.util.Random;
 import java.util.Set;
-import org.bukkit.World;
-import org.bukkit.entity.Arrow;
-import org.bukkit.entity.Entity;
-import org.bukkit.entity.LivingEntity;
-import org.bukkit.entity.Projectile;
+import org.bukkit.entity.*;
 import org.bukkit.event.EventHandler;
 import org.bukkit.event.Listener;
 import org.bukkit.event.entity.EntityDamageByEntityEvent;
@@ -25,13 +21,13 @@ import tc.oc.world.NMSHacks;
 
 @ListenerScope(MatchScope.RUNNING)
 public class ModifyBowProjectileMatchModule extends MatchModule implements Listener {
-  protected final Class<? extends Entity> cls;
+  protected final Class<? extends Projectile> cls;
   protected final float velocityMod;
   protected final Set<PotionEffect> potionEffects;
   protected final Random random = new Random();
 
   public ModifyBowProjectileMatchModule(
-      Match match, Class<? extends Entity> cls, float velocityMod, Set<PotionEffect> effects) {
+      Match match, Class<? extends Projectile> cls, float velocityMod, Set<PotionEffect> effects) {
     super(match);
     this.cls = cls;
     this.velocityMod = velocityMod;
@@ -40,32 +36,26 @@ public class ModifyBowProjectileMatchModule extends MatchModule implements Liste
 
   @EventHandler(ignoreCancelled = true)
   public void changeBowProjectile(EntityShootBowEvent event) {
+    Projectile oldProjectile = (Projectile) event.getProjectile();
+    Entity newProjectile = event.getProjectile();
     Plugin plugin = PGM.get();
-    Entity newProjectile;
 
-    if (this.cls == Arrow.class && event.getProjectile() instanceof Arrow) {
-      // Don't change the projectile if it's an Arrow and the custom entity type is also Arrow
-      newProjectile = event.getProjectile();
-    } else {
-      // Replace the projectile
-      World world = event.getEntity().getWorld();
-      Projectile oldEntity = (Projectile) event.getProjectile();
-      newProjectile = world.spawn(oldEntity.getLocation(), this.cls);
+    // Change the projectile if the custom entity type is not "Arrow"
+    if (this.cls != Arrow.class) {
+      // Replace the old projectile
+      newProjectile = event.getEntity().launchProjectile(this.cls);
       event.setProjectile(newProjectile);
 
-      // Copy some things from the old projectile
-      newProjectile.setVelocity(oldEntity.getVelocity());
-      newProjectile.setFallDistance(oldEntity.getFallDistance());
-      newProjectile.setFireTicks(oldEntity.getFireTicks());
-
-      if (newProjectile instanceof Projectile) {
-        ((Projectile) newProjectile).setShooter(oldEntity.getShooter());
-        ((Projectile) newProjectile).setBounce(oldEntity.doesBounce());
-      }
+      // Copy properties from old projectile
+      newProjectile.setVelocity(oldProjectile.getVelocity());
+      newProjectile.setFallDistance(oldProjectile.getFallDistance());
+      newProjectile.setFireTicks(oldProjectile.getFireTicks());
+      ((Projectile) newProjectile).setShooter(oldProjectile.getShooter());
+      ((Projectile) newProjectile).setBounce(oldProjectile.doesBounce());
 
       // Save some special properties of Arrows
-      if (oldEntity instanceof Arrow) {
-        Arrow arrow = (Arrow) oldEntity;
+      if (oldProjectile instanceof Arrow) {
+        Arrow arrow = (Arrow) oldProjectile;
         newProjectile.setMetadata("critical", new FixedMetadataValue(plugin, arrow.isCritical()));
         newProjectile.setMetadata(
             "knockback", new FixedMetadataValue(plugin, arrow.getKnockbackStrength()));
@@ -76,7 +66,6 @@ public class ModifyBowProjectileMatchModule extends MatchModule implements Liste
 
     // Tag the projectile as custom
     newProjectile.setMetadata("customProjectile", new FixedMetadataValue(plugin, true));
-
     getMatch().callEvent(new EntityLaunchEvent(newProjectile, event.getEntity()));
   }
 

--- a/src/main/java/tc/oc/pgm/modules/ModifyBowProjectileModule.java
+++ b/src/main/java/tc/oc/pgm/modules/ModifyBowProjectileModule.java
@@ -4,7 +4,7 @@ import java.util.HashSet;
 import java.util.Set;
 import java.util.logging.Logger;
 import org.bukkit.entity.Arrow;
-import org.bukkit.entity.Entity;
+import org.bukkit.entity.Projectile;
 import org.bukkit.potion.PotionEffect;
 import org.jdom2.Document;
 import org.jdom2.Element;
@@ -19,12 +19,12 @@ import tc.oc.xml.InvalidXMLException;
 
 @ModuleDescription(name = "Modify Bow Projectile")
 public class ModifyBowProjectileModule extends MapModule {
-  protected final Class<? extends Entity> cls;
+  protected final Class<? extends Projectile> cls;
   protected final float velocityMod;
   protected final Set<PotionEffect> potionEffects;
 
   public ModifyBowProjectileModule(
-      Class<? extends Entity> cls, float velocityMod, Set<PotionEffect> effects) {
+      Class<? extends Projectile> cls, float velocityMod, Set<PotionEffect> effects) {
     this.cls = cls;
     this.velocityMod = velocityMod;
     potionEffects = effects;
@@ -43,7 +43,7 @@ public class ModifyBowProjectileModule extends MapModule {
   public static ModifyBowProjectileModule parse(
       MapModuleContext context, Logger logger, Document doc) throws InvalidXMLException {
     boolean changed = false;
-    Class<? extends Entity> projectile = Arrow.class;
+    Class<? extends Projectile> projectile = Arrow.class;
     float velocityMod = 1;
     Set<PotionEffect> potionEffects = new HashSet<>();
 
@@ -56,7 +56,7 @@ public class ModifyBowProjectileModule extends MapModule {
 
       Element projectileElement = parent.getChild("projectile");
       if (projectileElement != null) {
-        projectile = XMLUtils.parseEntityType(projectileElement);
+        projectile = XMLUtils.parseProjectileType(projectileElement);
         changed = true;
       }
 

--- a/src/main/java/tc/oc/pgm/projectile/ProjectileModule.java
+++ b/src/main/java/tc/oc/pgm/projectile/ProjectileModule.java
@@ -57,7 +57,7 @@ public class ProjectileModule extends MapModule {
               "click action",
               ClickAction.BOTH);
       Class<? extends Entity> entity =
-          XMLUtils.parseEntityTypeAttribute(projectileElement, "projectile", Arrow.class);
+          XMLUtils.parseProjectileTypeAttribute(projectileElement, "projectile", Arrow.class);
       List<PotionEffect> potionKit = kitParser.parsePotions(projectileElement);
       Filter destroyFilter = filterParser.parseFilterProperty(projectileElement, "destroy-filter");
       Duration coolDown = XMLUtils.parseDuration(projectileElement.getAttribute("cooldown"));

--- a/src/main/java/tc/oc/pgm/util/XMLUtils.java
+++ b/src/main/java/tc/oc/pgm/util/XMLUtils.java
@@ -19,7 +19,7 @@ import org.apache.commons.codec.binary.Base64;
 import org.bukkit.*;
 import org.bukkit.attribute.AttributeModifier;
 import org.bukkit.enchantments.Enchantment;
-import org.bukkit.entity.Entity;
+import org.bukkit.entity.Projectile;
 import org.bukkit.material.MaterialData;
 import org.bukkit.potion.PotionEffect;
 import org.bukkit.potion.PotionEffectType;
@@ -498,31 +498,34 @@ public class XMLUtils {
     return node == null ? def : parseSecondDuration(node);
   }
 
-  public static Class<? extends Entity> parseEntityType(Element el) throws InvalidXMLException {
-    return parseEntityType(new Node(el));
+  public static Class<? extends Projectile> parseProjectileType(Element el)
+      throws InvalidXMLException {
+    return parseProjectileType(new Node(el));
   }
 
-  public static Class<? extends Entity> parseEntityTypeAttribute(
-      Element el, String attributeName, Class<? extends Entity> def) throws InvalidXMLException {
-    Node node = Node.fromAttr(el, attributeName);
-    return node == null ? def : parseEntityType(node);
+  public static Class<? extends Projectile> parseProjectileType(Node node)
+      throws InvalidXMLException {
+    return parseProjectileType(node, node.getValue());
   }
 
-  public static Class<? extends Entity> parseEntityType(Node node) throws InvalidXMLException {
-    return parseEntityType(node, node.getValue());
-  }
-
-  public static Class<? extends Entity> parseEntityType(Node node, String value)
+  public static Class<? extends Projectile> parseProjectileType(Node node, String value)
       throws InvalidXMLException {
     if (!value.matches("[a-zA-Z0-9_]+")) {
-      throw new InvalidXMLException("Invalid entity type '" + value + "'", node);
+      throw new InvalidXMLException("Invalid projectile type '" + value + "'", node);
     }
 
     try {
-      return Class.forName("org.bukkit.entity." + value).asSubclass(Entity.class);
+      return Class.forName("org.bukkit.entity." + value).asSubclass(Projectile.class);
     } catch (ClassNotFoundException | ClassCastException e) {
-      throw new InvalidXMLException("Invalid entity type '" + value + "'", node);
+      throw new InvalidXMLException("Invalid projectile type '" + value + "'", node);
     }
+  }
+
+  public static Class<? extends Projectile> parseProjectileTypeAttribute(
+      Element el, String attributeName, Class<? extends Projectile> def)
+      throws InvalidXMLException {
+    Node node = Node.fromAttr(el, attributeName);
+    return node == null ? def : parseProjectileType(node);
   }
 
   public static Vector parseVector(Node node, String value) throws InvalidXMLException {

